assets/mdk/proof-dependencies.mdk

```diff
@@ -0,0 +1,87 @@
+Colorizer     : dafnyx
+
+Heading Depth: 0
+
+Css Header:
+  body.madoko {
+    max-width: 100%;
+    margin: 0;
+    padding: 0;
+  }
+  .page-content {
+    padding: 0;
+  }
+
+# Introduction
+
+A verified program is only as good as its specification. If the properties you want to prove about your program are contradictory, any program will satisfy them. If the properties you want to prove don't cover all of the program's behavior, any implementation will satisfy the uncovered portions. Specification problems of these sort result in verification that is technically sound, but that may provide a false sense of security that the program behaves as intended.
+
+With v4.4.0[^1] Dafny will include features to detect problems of this sort, using a notion of _proof dependency_. A proof dependency is a portion of a program that needed to be taken into account in order to construct a proof that the program is correct.
+
+[^1]: Or any nightly build after 2023-10-24.
+
+If it's possible to prove a program correct without taking the final proof goal (_e.g._, `assert` statement or `ensures` clause) into account, that goal must have been proven using contradictory assumptions or infeasible branches in the program. Contradictory specifications are particularly easy to construct by accident when writing assumed specifications about external or yet-to-be-written code. Consider the following example.
```

Comment:
I don't understand: "If it's possible to prove a program correct without taking the final proof goal (_e.g._, `assert` statement or `ensures` clause) into account, that goal must have been proven using contradictory assumptions or infeasible branches in the program." what is the program correctness we're talking about? Also, couldn't the goal just be useless? Sometimes it does happen that we provide more assert than strictly necessary for clarity. I think I am very confused by this paragraph.

Reply:
I'll try to clarify this. A useless assertion would show up in a different way.